drivers/gles2/shaders/stdlib.glsl

```diff
@@ -51,14 +183,163 @@ highp vec3 round(highp vec3 x) {
+
+highp vec4 roundEven(highp vec4 x) {
+	return vec4(roundEven(x.x), roundEven(x.y), roundEven(x.z), roundEven(x.w));
+}
+
+#endif
+
+#ifdef IS_INF_USED
+
+bool isinf(highp float x) {
+	return (2 * x == x) && (x != 0);
+}
+
+bvec2 isinf(highp vec2 x) {
+	return bvec2((2 * x.x == x.x) && (x.x != 0), (2 * x.y == x.y) && (x.y != 0));
+}
+
+bvec3 isinf(highp vec3 x) {
+	return bvec3((2 * x.x == x.x) && (x.x != 0), (2 * x.y == x.y) && (x.y != 0), (2 * x.z == x.z) && (x.z != 0));
+}
+
+bvec4 isinf(highp vec4 x) {
+	return bvec4((2 * x.x == x.x) && (x.x != 0), (2 * x.y == x.y) && (x.y != 0), (2 * x.z == x.z) && (x.z != 0), (2 * x.w == x.w) && (x.w != 0));
+}
+
+#endif
+
+#ifdef IS_NAN_USED
+
+bool isnan(highp float x) {
+	return x != x;
+}
+
+bvec2 isnan(highp vec2 x) {
+	return bvec2(x.x != x.x, x.y != x.y);
+}
+
+bvec3 isnan(highp vec3 x) {
+	return bvec3(x.x != x.x, x.y != x.y, x.z != x.z);
+}
+
+bvec4 isnan(highp vec4 x) {
+	return bvec4(x.x != x.x, x.y != x.y, x.z != x.z, x.w != x.w);
+}
+
+#endif
+
+#ifdef TRUNC_USED
+
+highp float trunc(highp float x) {
+	return x < 0 ? -floor(-x) : floor(x);
+}
+
+highp vec2 trunc(highp vec2 x) {
+	return vec2(x.x < 0 ? -floor(-x.x) : floor(x.x), x.y < 0 ? -floor(-x.y) : floor(x.y));
+}
+
+highp vec3 trunc(highp vec3 x) {
+	return vec3(x.x < 0 ? -floor(-x.x) : floor(x.x), x.y < 0 ? -floor(-x.y) : floor(x.y), x.z < 0 ? -floor(-x.z) : floor(x.z));
+}
+
+highp vec4 trunc(highp vec4 x) {
+	return vec4(x.x < 0 ? -floor(-x.x) : floor(x.x), x.y < 0 ? -floor(-x.y) : floor(x.y), x.z < 0 ? -floor(-x.z) : floor(x.z), x.w < 0 ? -floor(-x.w) : floor(x.w));
+}
+
+#endif
+
+#ifdef DETERMINANT_USED
+
+highp float determinant(highp mat2 m) {
+	return m[0].x * m[1].y - m[1].x * m[0].y;
+}
+
+highp float determinant(highp mat3 m) {
+	return m[0].x * (m[1].y * m[2].z - m[2].y * m[1].z) - m[1].x * (m[0].y * m[2].z - m[2].y * m[0].z) + m[2].x * (m[0].y * m[1].z - m[1].y * m[0].z);
+}
+
+highp float determinant(highp mat4 m) {
+	highp float s00 = m[2].z * m[3].w - m[3].z * m[2].w;
+	highp float s01 = m[2].y * m[3].w - m[3].y * m[2].w;
+	highp float s02 = m[2].y * m[3].z - m[3].y * m[2].z;
+	highp float s03 = m[2].x * m[3].w - m[3].x * m[2].w;
+	highp float s04 = m[2].x * m[3].z - m[3].x * m[2].z;
+	highp float s05 = m[2].x * m[3].y - m[3].x * m[2].y;
+	highp vec4 c = vec4((m[1].y * s00 - m[1].z * s01 + m[1].w * s02), -(m[1].x * s00 - m[1].z * s03 + m[1].w * s04), (m[1].x * s01 - m[1].y * s03 + m[1].w * s05), -(m[1].x * s02 - m[1].y * s04 + m[1].z * s05));
+	return m[0].x * c.x + m[0].y * c.y + m[0].z * c.z + m[0].w * c.w;
+}
+
 #endif
 
 #ifndef USE_GLES_OVER_GL
-highp mat4 transpose(highp mat4 src) {
+
+#ifdef TRANSPOSE_USED
+
+highp mat2 transpose(highp mat2 m) {
+	return mat2(
+			vec2(m[0].x, m[1].x),
+			vec2(m[0].y, m[1].y));
+}
+
+highp mat3 transpose(highp mat3 m) {
+	return mat3(
+			vec3(m[0].x, m[1].x, m[2].x),
+			vec3(m[0].y, m[1].y, m[2].y),
+			vec3(m[0].z, m[1].z, m[2].z));
+}
+
+highp mat4 transpose(highp mat4 m) {
```

Comment:
``transpose(mat4)`` is used inside canvas.glsl and scene.glsl, so it should be exposed outside the ``#ifdef TRANSPOSE_USED`` but inside the ``#ifndef GLES_OVER_GL``. Sorry I missed this before.

Reply:
No problem - fixed